AOTAutograd should generate same hash on different processes with identical inputs

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #170683

Mostly done by Claude. 
When we call hash on dynamo sources that contains strings, we can't guarantee stability as python hash() function returns different hash on each processes. In Dynamo, there is a logic to cache the hash which is causing problem because now we are storing the hash instead of computing on the fly. The solution is to opt into stable hashing. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

tugsbayasgalan committed:
SAOTAutograd should generate same hash on different processes with identical inputs

[ghstack-poisoned]

bobrenjc93 commented:
Reading the PR more carefully it seems like an unstable source hash causes the `aot_autograd_arg_pos_to_source` of `AOTConfig` to change for the same source. Is that correct? I'm really surprised AOTAutogradCache works at the 70+% hit rate if so...

tugsbayasgalan committed:
Update on "AOTAutograd should generate same hash on different processes with identical inputs"


Mostly done by Claude. 
When we call hash on dynamo sources that contains strings, we can't guarantee stability as python hash() function returns different hash on each processes. In Dynamo, there is a logic to cache the hash which is causing problem because now we are storing the hash instead of computing on the fly. The solution is to exclude hash when pickling. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "AOTAutograd should generate same hash on different processes with identical inputs"


Mostly done by Claude. 
When we call hash on dynamo sources that contains strings, we can't guarantee stability as python hash() function returns different hash on each processes. In Dynamo, there is a logic to cache the hash which is causing problem because now we are storing the hash instead of computing on the fly. The solution is to opt into stable hashing. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge -i

ezyang commented on torch/_guards.py:
```diff
@@ -1140,13 +1151,24 @@ def dataclass_with_cached_hash(
+        def __hash__(self: Any) -> int:
             if not hasattr(self, "_hash"):
-                object.__setattr__(self, "_hash", old_hash(self))
+                object.__setattr__(self, "_hash", _stable_hash(self))
```
I think I would kind of like to know why `__hash__` is being used for a cross-process hash. It shouldn't be; you should be using some other API in this case.
